File: src/infinicore/spmat.cc
#include "infinicore/spmat.hpp"
#include "utils.hpp"

namespace infinicore {

SpMat SpMat::csr(Tensor crow_indices, Tensor col_indices, Tensor values, Size rows, Size cols) {
    return SpMat{std::make_shared<SpMatImpl>(crow_indices, col_indices, values, rows, cols)};
}

SpMatImpl *SpMat::operator->() {
    return impl_.get();
}

const SpMatImpl *SpMat::operator->() const {
    return impl_.get();
}

SpMat::operator bool() const {
    return impl_ != nullptr;
}

SpMatImpl::SpMatImpl(Tensor crow_indices, Tensor col_indices, Tensor values, Size rows, Size cols)
    : crow_indices_(crow_indices),
      col_indices_(col_indices),
      values_(values),
      rows_(rows),
      cols_(cols),
      desc_(nullptr) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(crow_indices_, col_indices_, values_);
    INFINICORE_ASSERT(crow_indices_->is_contiguous());
    INFINICORE_ASSERT(col_indices_->is_contiguous());
    INFINICORE_ASSERT(values_->is_contiguous());
    INFINICORE_CHECK_ERROR(infiniopCreateCsrSpMatDescriptor(
        &desc_,
        rows_,
        cols_,
        values_->numel(),
        values_->desc(),
        crow_indices_->desc(),
        col_indices_->desc(),
        values_->data(),
        crow_indices_->data(),
        col_indices_->data()));
}

SpMatImpl::~SpMatImpl() {
    if (desc_) {
        infiniopDestroySpMatDescriptor(desc_);
        desc_ = nullptr;
    }
}

Size SpMatImpl::rows() const {
    return rows_;
}

Size SpMatImpl::cols() const {
    return cols_;
}

Size SpMatImpl::nnz() const {
    return values_->numel();
}

DataType SpMatImpl::dtype() const {
    return values_->dtype();
}

DataType SpMatImpl::index_dtype() const {
    return crow_indices_->dtype();
}

Device SpMatImpl::device() const {
    return values_->device();
}

const Tensor &SpMatImpl::crow_indices() const {
    return crow_indices_;
}

const Tensor &SpMatImpl::col_indices() const {
    return col_indices_;
}

const Tensor &SpMatImpl::values() const {
    return values_;
}

infiniopSpMatDescriptor_t SpMatImpl::desc() const {
    return desc_;
}

} // namespace infinicore

File: python/infinicore/spmat.py
import infinicore.device
import infinicore.dtype
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


class SpMat:
    _underlying: _infinicore.SpMat

    def __init__(self, underlying):
        self._underlying = underlying

    @property
    def rows(self):
        return self._underlying.rows

    @property
    def cols(self):
        return self._underlying.cols

    @property
    def nnz(self):
        return self._underlying.nnz

    @property
    def shape(self):
        return [self.rows, self.cols]

    @property
    def dtype(self):
        return infinicore.dtype.dtype(self._underlying.dtype)

    @property
    def index_dtype(self):
        return infinicore.dtype.dtype(self._underlying.index_dtype)

    @property
    def device(self):
        return infinicore.device._from_infinicore_device(self._underlying.device)

    @property
    def crow_indices(self):
        return Tensor(self._underlying.crow_indices)

    @property
    def col_indices(self):
        return Tensor(self._underlying.col_indices)

    @property
    def values(self):
        return Tensor(self._underlying.values)

    def __matmul__(self, other):
        if other.ndim == 1:
            from infinicore.ops.spmv import spmv

            return spmv(self, other)

        from infinicore.ops.spmm import spmm

        return spmm(self, other)


def csr_spmat(crow_indices, col_indices, values, size):
    if len(size) != 2:
        raise ValueError("CSR sparse matrix size must be a 2-tuple/list")
    return SpMat(
        _infinicore.csr_spmat(
            crow_indices._underlying,
            col_indices._underlying,
            values._underlying,
            size[0],
            size[1],
        )
    )
